plugin/core/promise.py

```diff
@@ -126,26 +134,26 @@ def recheck_resolve_status(_: Any) -> None:
             return Promise(handler)
         return Promise.resolve()
 
-    def __init__(self, fullfill_func: FullfillFunc) -> None:
+    def __init__(self, executor_func: ExecutorFunc[T]) -> None:
         """Initialize Promise object.
 
         Arguments:
-            fullfill_func: A function that is executed immediately by this Promise.
+            executor_func: A function that is executed immediately by this Promise.
             It gets passed a "resolve" function. The "resolve" function, when
             called, resolves the Promise with the value passed to it.
         """
-        self.value = None  # type: Any
+        self.value = None  # type: T
         self.resolved = False
         self.mutex = threading.Lock()
-        self.callbacks = []  # type: List[ResolveFunc]
-        fullfill_func(lambda value=None: self._do_resolve(value))
+        self.callbacks = []  # type: List[ResolveFunc[T]]
+        executor_func(lambda value=None: self._do_resolve(value))
 
     def __repr__(self) -> str:
         if self.resolved:
             return 'Promise({})'.format(self.value)
         return 'Promise(<pending>)'
 
-    def then(self, callback: ThenFunc) -> 'Promise':
+    def then(self, onfullfilled: FullfillFunc[T, TResult]) -> 'Promise[TResult]':
```

Comment:
This doesn't seem right. And this place is probably the hardest to get right because it's the monadic join operation.

1. When you pass a `Callable[[T], TResult]` in here, then the return type of this method is `Promise[TResult]`.
2. When you pass a `Callable[[T], Promise[TResult]` in here, then the return type is **also** `Promise[TResult]`.

Point (2) is not taken care of here I think.

Reply:
Oh never mind, it is taken care of :)

But not really in a nice way. Maybe we can make use of `@overload` here?

Reply:
Something like

```python
@overload
def then(self, f: Callable[[T], TResult]) -> Promise[TResult]: ...
@overload
def then(self, f: Callable[[T], Promise[TResult]]) -> Promise[TResult]: ...

def then(self, f: Callable[[T], Union[TResult, Promise[TResult], None]]) -> Promise:
    # the actual implementation
```

By the way, awesome that you're trying to tackle this!

Reply:
`@overload` is new in 3.8, as far as I can see. So won't work here.

But what is not nice about it?

Reply:
Looks like I'm wrong, everything works fine here :)

Reply:
I admit types get quite crazy when trying to wrap this around your head. :)